modules/data-streams/src/internalClusterTest/java/org/elasticsearch/datastreams/logsdb/qa/AbstractChallengeTest.java

```diff
@@ -0,0 +1,281 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.datastreams.logsdb.qa;
+
+import org.elasticsearch.action.DocWriteRequest;
+import org.elasticsearch.action.admin.indices.refresh.RefreshRequest;
+import org.elasticsearch.action.admin.indices.template.put.TransportPutComposableIndexTemplateAction;
+import org.elasticsearch.action.bulk.BulkRequest;
+import org.elasticsearch.action.bulk.BulkResponse;
+import org.elasticsearch.action.datastreams.CreateDataStreamAction;
+import org.elasticsearch.action.index.IndexRequest;
+import org.elasticsearch.action.search.SearchRequest;
+import org.elasticsearch.action.search.SearchResponse;
+import org.elasticsearch.action.support.broadcast.BroadcastResponse;
+import org.elasticsearch.action.support.master.AcknowledgedResponse;
+import org.elasticsearch.cluster.metadata.ComposableIndexTemplate;
+import org.elasticsearch.cluster.metadata.Template;
+import org.elasticsearch.common.CheckedSupplier;
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.compress.CompressedXContent;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.core.CheckedConsumer;
+import org.elasticsearch.core.Tuple;
+import org.elasticsearch.datastreams.DataStreamsPlugin;
+import org.elasticsearch.plugins.Plugin;
+import org.elasticsearch.search.builder.SearchSourceBuilder;
+import org.elasticsearch.test.ESSingleNodeTestCase;
+import org.elasticsearch.xcontent.XContentBuilder;
+import org.elasticsearch.xcontent.XContentFactory;
+import org.junit.Before;
+
+import java.io.IOException;
+import java.util.Collection;
+import java.util.List;
+import java.util.function.Supplier;
+
+/**
+ * This class encapsulates the logic for running tests using two data streams to compare results
+ * from indexing and querying operations. The `oracle` data stream serves as the source of truth,
+ * while the `challenge` data stream is tested against the oracle. Comparing operation results
+ * on both data streams ensures that the challenge data stream behaves as expected relative to the
+ * oracle data stream.
+ */
+public abstract class AbstractChallengeTest extends ESSingleNodeTestCase {
```

Comment:
What is the reason for choosing `ESSingleNodeTestCase` here? Ideally I think we would want to do quality assurance via a java rest integration test?

Reply:
I will try to see if we can have both...having a different base class and sharing what we can share. The base class to use is important only when it comes to the APIs we call because it is at higher abstraction level. And of course difference in the node we start.

Reply:
Using `ESRestTest` makes matching results a bit more involved since we have to parse the whole HTTP/Json response. Using `ESSingleNodeTestCase` allows us to use internal objects like `InternalAggregation`, `SearchResponse` and so on which would allow us to ignore differences in output format representation (like json). The question is...when checking the reponse should we check the actual result of the query or its (json) representation? I honestly don't think here we have to test the output format. As a result I like it more how it was before the last commit.

Reply:
Right, we should be able to compare results between standard and logsdb. In case of rest tests response is basically a map. But the test cluster is more realistic now and closer to what users run with it. For example the stack plugin is active and we should run with default templates as well.